pkg/azure/provider/provider.go

```diff
@@ -0,0 +1,195 @@
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("Failed to extract VM names from VMs and NICs for resourceGroup: %s, Err: %v", providerSpec.ResourceGroup, err))
+		return
+	}
+	resp = helpers.ConstructMachineListResponse(providerSpec.Location, vmNames)
+	return
+}
+
+func (d defaultDriver) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (resp *driver.CreateMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, createMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	vmName := req.Machine.Name
+	nicName := utils.CreateNICName(vmName)
+
+	imageReference, plan, err := helpers.ProcessVMImageConfiguration(ctx, d.factory, connectConfig, providerSpec, vmName)
+	if err != nil {
+		return
+	}
+	subnet, err := helpers.GetSubnet(ctx, d.factory, connectConfig, providerSpec)
+	if err != nil {
+		return
+	}
+
+	nicID, err := helpers.CreateNICIfNotExists(ctx, d.factory, connectConfig, providerSpec, subnet, nicName)
+	if err != nil {
+		return
+	}
+
+	vm, err := helpers.CreateOrUpdateVM(ctx, d.factory, connectConfig, providerSpec, imageReference, plan, req.Secret, nicID, vmName)
+	if err != nil {
+		return
+	}
+	resp = helpers.ConstructCreateMachineResponse(providerSpec.Location, vmName)
+	helpers.LogVMCreation(providerSpec.Location, providerSpec.ResourceGroup, vm)
+	return
+}
+
+func (d defaultDriver) DeleteMachine(ctx context.Context, req *driver.DeleteMachineRequest) (resp *driver.DeleteMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, deleteMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	var (
+		resourceGroup = providerSpec.ResourceGroup
+		vmName        = strings.ToLower(req.Machine.Name)
+	)
+	// Check if Deletion of the machine (VM, NIC, Disks) can be completely skipped.
+	skipDelete, err := helpers.SkipDeleteMachine(ctx, d.factory, connectConfig, resourceGroup)
+	if err != nil {
+		return
+	}
+	if skipDelete {
+		klog.Warningf("skipping delete of Machine [ResourceGroup: %s, Name: %s] since the ResourceGroup no longer exists", resourceGroup, req.Machine.Name)
+		resp = &driver.DeleteMachineResponse{}
+		return
+	}
+
+	vmAccess, err := d.factory.GetVirtualMachinesAccess(connectConfig)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("failed to create virtual machine access to process request: [resourceGroup: %s, vmName: %s], Err: %v\n", resourceGroup, vmName, err))
+		return
+	}
+	vm, err := clienthelpers.GetVirtualMachine(ctx, vmAccess, resourceGroup, vmName)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("failed to get virtual machine for VM: [resourceGroup: %s, name: %s], Err: %v", resourceGroup, vmName, err))
+		return
+	}
+	if vm == nil {
```

Comment:
could you add a comment to mention for which cases this individual delete of nics and disks is being done?

Reply:
done